engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/age/AbstractAgeInstantBasedValidator.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.constraintvalidators.hv.age;
+
+import java.lang.annotation.Annotation;
+import java.lang.invoke.MethodHandles;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+
+import javax.validation.ConstraintValidatorContext;
+
+import org.hibernate.validator.constraintvalidation.HibernateConstraintValidator;
+import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorInitializationContext;
+import org.hibernate.validator.internal.util.logging.Log;
+import org.hibernate.validator.internal.util.logging.LoggerFactory;
+
+/**
+ * Base class for all age validators that use an {@link Instant} to be compared to the age reference.
+ *
+ * @author Hillmer Chona
+ * @since 6.0.8
+ */
+public abstract class AbstractAgeInstantBasedValidator<C extends Annotation, T>
+		implements HibernateConstraintValidator<C, T> {
+
+	private static final Log LOG = LoggerFactory.make( MethodHandles.lookup() );
+
+	private Clock referenceClock;
+
+	protected int referenceAge;
+
+	protected boolean inclusive;
+
+	protected ChronoUnit unit;
+
+	public void initialize(
+			int referenceAge,
+			ChronoUnit unit,
+			boolean inclusive,
+			HibernateConstraintValidatorInitializationContext initializationContext) {
+		try {
+			this.referenceClock = Clock.offset(
+					initializationContext.getClockProvider().getClock(),
+					getEffectiveTemporalValidationTolerance( initializationContext.getTemporalValidationTolerance() )
+			);
+		}
+		catch (Exception e) {
+			throw LOG.getUnableToGetCurrentTimeFromClockProvider( e );
+		}
+		this.referenceAge = referenceAge;
+		this.unit = unit;
+		this.inclusive = inclusive;
+	}
+
+	@Override
+	public boolean isValid(T value, ConstraintValidatorContext context) {
+		// null values are valid
+		if ( value == null ) {
+			return true;
+		}
+		// As Instant does not support plus operation on ChronoUnits greater than DAYS we need to convert it to LocalDate
+		// first, which supports such operations.
+
+		long result = getInstant( value ).atZone( ZoneOffset.ofHours( 0 ) )
```

Comment:
as mentioned in another comment, I think we should stick to `LocalDateTime` so there will be no errors for different units. In case we use `LocalDate` here `UnsupportedTemporalTypeException` will be thrown for minutes, hours and other time units.

Reply:
@marko-bekhta for InstanteBase I don't see any problem using either LocalDate or LocalDateTime, but it is necessary convert to LocalDate before comparing Calendar, the reason is that Calendar has time, even though if you were born for example at 10:00 pm your new age  starts from  00:00 AM.

    long result = getInstant( value ).atZone( ZoneOffset.ofHours( 0 ) )
    .toLocalDateTime().plus( referenceAge, unit ).toLocalDate()
    .compareTo( LocalDateTime.now( referenceClock).toLocalDate() );


For TimeBased, I got an `UnsupportedTemporalTypeException` yesterday with Year,  as you say with LocalDate. 

A possible solution can be to create a method LocalDateTime getReferenceValue(T value, Clock reference), and for Year return the year that comes on value plus January 1st 00:00 AM, For LocalDate the date that comes on value plus 00:00 AM. I mean they only return a LocalDateTime, adding the values (date and/or time) to make the T value consistent . But I'm not sure about it. Can you see other solution?

Reply:
@Hilmerc that's a very good point on the time part of birthday and how we calculate the age. Thinking more about it, it might be that we wanted to do to much with one constraint and we might slightly moved away from your original intention for it. Seems that there are next two ways how it can move forward from this point:

1. As we already have the `unit` in place and it is possible to specify age for example in minutes it wouldn't make sense to truncate all that data up to days which is suitable for calculating the age of a person. Hence if we want to keep `unit` attribute in place and also have a correct logic for calculating an age of a person we would need to add another attribute to this constraint. Something like `ChronoUnit truncateTo() default ChronoUnit.NANOS;` a time unit to which we want to truncate the date-time. In such case if we want an age of a person we would use `truncateTo=ChronoUnit.DAYS`. And if we would like to keep the time part we would leave it as default.
2. Ignore the idea of having the ability to specify units for an age and keep this particular constraint as simple as an age of a person. But as it looks to be an interesting and useful case to be able to specify constraints on how much time have passed think of a different constraint for it.

Personally, I'm leaning more towards (1), as it will be a more versatile constraint. If you like this approach and would want to give it a try we'd need to add this additional attribute to a constraint and then use `LocalDateTime#truncatedTo(TemporalUnit)` in validators. Also as we would only operate with `LocalDateTime` instances in the validator to prevent exceptions on plus/minus operations maybe it will actually make sense to have a single base validator for all types? We could then modify the current instant one to something like:
```java
class AbstractAgeValidator {
....

    public void initialize(
        int referenceAge,
        ChronoUnit unit,
        boolean inclusive,
        HibernateConstraintValidatorInitializationContext initializationContext) {
    ....
    }

    @Override
    public boolean isValid(T value, ConstraintValidatorContext context) {
    ....
    }
    protected abstract LocalDateTime getLocalDateTime(T value);
    ....
}

class AbstractAgeInstantBasedValidator extends AbstractAgeValidator {
    protected abstract Instant getInstant(T value);

    @Override
    protected LocalDateTime getLocalDateTime(T value){
        return LocalDateTime.ofInstant(getInstant(value));
    }
}
```
and for types like `LocalDate` and others we could just use methods on them to get to `LocalDateTime` and return directly from `getLocalDateTime` method. 

Also, just as a note, in any of these cases, we would need to write up a better explanation of what the constraint is checking and how it will behave. Just to set the users expectations right :) But don't think about this part yet. That's something for later.

Reply:
@marko-bekhta after some work and tests I figured out that the attribute  ChronoUnit#truncateTo() can generate some problems like ChronoUnit#Unit.  

    jshell> LocalDateTime.now( ZoneOffset.UTC  ).truncatedTo(ChronoUnit.YEARS);
    |  java.time.temporal.UnsupportedTemporalTypeException thrown: Unit is too large to be used for truncation
    jshell> LocalDateTime.now( ZoneOffset.UTC  ).truncatedTo(ChronoUnit.WEEKS);
    |  java.time.temporal.UnsupportedTemporalTypeException thrown: Unit is too large to be used     for truncation

I think we should eliminate the ability to specify a unit, make validators with years and, after use and feedback, we can explore the possibility of returning to our idea of using Units.

WDYT?

Reply:
@Hilmerc thanks a lot for experimenting with this approach. So it seems, we wanted to much from one constraint. Let's remove the units then and clean things up to keep this constraint just for age in years. I still think that the above idea with one abstract validator for all types would work (Just use `LocalDate` instead of `LocalDateTime`).